# Treatment of data

This notebook is used to clean a dataset downloaded from kaggle with data on global shark attacks

## Table of contents
- Import libraries
- Import dataset
- Explore dataset through visualization
- Treat dataset 

## Import libraries

In [1]:
import pandas as pd
import numpy as np
import re

## Import dataset

In [2]:
data = pd.read_csv("../data/attacks.csv", encoding = "ISO-8859-1")

In [3]:
data

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

## Explore dataset through visualization

In [4]:
data.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

With this dataset, I would like to test three hypotheses. First, I hypothesize that the number of attacks have gone down throughout the years. Second, I hypothesize that the type of attack will be correlated with the sex of the victim. Finally, the activity that the victim was doing is correlated with the fatality of the attack. 

The first step is to identify all columns and try to identify available data, duplicated data and data that we don't understand

In [5]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

After understanding the available data, it is clear that there are some columns that are offering no new data. To confirm which columns can be deleted, it is also relevant to see how many rows are not just full of 'NaNs' so that we can know how many rows have actual data.

In [6]:
# this code identifies rows where all values are NaN and returns a view of a df with these rows

full_nan_indxs = data.index[data.isnull().all(1)]
nans_r = data.iloc[full_nan_indxs]
nans_r

Case Number Date  Year Type Country Area Location Activity Name Sex   \
8702          NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8703          NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8704          NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8705          NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8706          NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
...           ...  ...   ...  ...     ...  ...      ...      ...  ...  ...   
25717         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25718         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25719         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25720         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25721         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
8702   ...      NaN                    NaN  NaN          NaN  NaN   
8703   ...      NaN                    NaN  NaN          NaN  NaN   
8704   ...      NaN                    NaN  NaN          NaN  NaN   
8705   ...      NaN                    NaN  NaN          NaN  NaN   
8706   ...      NaN                    NaN  NaN          NaN  NaN   
...    ...      ...                    ...  ...          ...  ...   
25717  ...      NaN                    NaN  NaN          NaN  NaN   
25718  ...      NaN                    NaN  NaN          NaN  NaN   
25719  ...      NaN                    NaN  NaN          NaN  NaN   
25720  ...      NaN                    NaN  NaN          NaN  NaN   
25721  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
8702            NaN           NaN            NaN         NaN         NaN  
8703            NaN           NaN            NaN         NaN         NaN  
8704            NaN           NaN            NaN         NaN         NaN  
8705            NaN           NaN            NaN         NaN         NaN  
8706            NaN           NaN            NaN         NaN         NaN  
...             ...           ...            ...         ...         ...  
25717           NaN           NaN            NaN         NaN         NaN  
25718           NaN           NaN            NaN         NaN         NaN  
25719           NaN           NaN            NaN         NaN         NaN  
25720           NaN           NaN            NaN         NaN         NaN  
25721           NaN           NaN            NaN         NaN         NaN  

[17020 rows x 24 columns]

It looks like there are 17020 rows full of NaNs, so we should delete those later

In [7]:
# this code identifies columns with the highest percentage of NaNs 

data.isnull().sum().apply(lambda x: x/data.shape[0]).sort_values(ascending=False)

Unnamed: 22               0.999961
Unnamed: 23               0.999922
Time                      0.885394
Species                   0.865335
Age                       0.865062
Sex                       0.776970
Activity                  0.776154
Location                  0.775998
Fatal (Y/N)               0.775959
Area                      0.772694
Name                      0.763169
Country                   0.756949
Injury                    0.756094
Investigator or Source    0.755666
Type                      0.755161
Year                      0.755083
href formula              0.755044
Date                      0.755005
pdf                       0.755005
href                      0.755005
Case Number.1             0.755005
Case Number.2             0.755005
original order            0.754733
Case Number               0.661704
dtype: float64

Unnamed: 22 and 23 seem to be empty and could be potentially removed. Other columns with a high % of NaNs do contain relevant information (eg. time, species, age...) at there is a valid reason for not having values for all rows in these columns (this data does not always get collected). Lets look ak what Unnamed 22 and 23 contain.

In [8]:
data[data['Unnamed: 22'].notnull()]

Case Number         Date    Year        Type Country    Area  \
1478  2006.05.27  27-May-2006  2006.0  Unprovoked     USA  Hawaii   

                Location Activity          Name Sex   ... Species   \
1478  North Shore, O'ahu  Surfing  Bret Desmond    M  ...      NaN   

     Investigator or Source                     pdf  \
1478             R. Collier  2006.05.27-Desmond.pdf   

                                           href formula  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.05.27   

     Case Number.2 original order   Unnamed: 22 Unnamed: 23  
1478    2006.05.27         4825.0  stopped here         NaN  

[1 rows x 24 columns]

In [9]:
data[data['Unnamed: 23'].notnull()]

Case Number                  Date    Year        Type  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   Country         Area                    Location  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             Activity              Name Sex   \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      ...                                           Species   \
4415  ...  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840  ...                                                NaN   

                                 Investigator or Source  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href formula  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     Case Number.2 original order Unnamed: 22      Unnamed: 23  
4415    1952.03.30         1888.0         NaN           Teramo  
5840  1878.09.14.R          463.0         NaN  change filename  

[2 rows x 24 columns]

These columns contain irrelevant information and could therefore be removed to make the data frame cleaner

Now, following the order of columns, lets look at case number. This should be a unique identifier for each attack, but in fact it just gives the exact same data as the date column just beside it. Since shark attacks are relatively rare, the date could be unique for each attack, but seeing as it doesn't give more information than the date column, checking if the dates are unique is not worth it. The only thing worth checking is seeing if any rows  have NaN for date but have a case number.

In [10]:
data[data['Case Number'].notnull() & data.Date.isnull()]

Case Number Date  Year Type Country Area Location Activity Name Sex   \
6302            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6303            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6304            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6305            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6306            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
...           ...  ...   ...  ...     ...  ...      ...      ...  ...  ...   
8698            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8699            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8700            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8701            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
6302   ...      NaN                    NaN  NaN          NaN  NaN   
6303   ...      NaN                    NaN  NaN          NaN  NaN   
6304   ...      NaN                    NaN  NaN          NaN  NaN   
6305   ...      NaN                    NaN  NaN          NaN  NaN   
6306   ...      NaN                    NaN  NaN          NaN  NaN   
...    ...      ...                    ...  ...          ...  ...   
8698   ...      NaN                    NaN  NaN          NaN  NaN   
8699   ...      NaN                    NaN  NaN          NaN  NaN   
8700   ...      NaN                    NaN  NaN          NaN  NaN   
8701   ...      NaN                    NaN  NaN          NaN  NaN   
25722  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
6302            NaN           NaN         6304.0         NaN         NaN  
6303            NaN           NaN         6305.0         NaN         NaN  
6304            NaN           NaN         6306.0         NaN         NaN  
6305            NaN           NaN         6307.0         NaN         NaN  
6306            NaN           NaN         6308.0         NaN         NaN  
...             ...           ...            ...         ...         ...  
8698            NaN           NaN            NaN         NaN         NaN  
8699            NaN           NaN            NaN         NaN         NaN  
8700            NaN           NaN            NaN         NaN         NaN  
8701            NaN           NaN            NaN         NaN         NaN  
25722           NaN           NaN            NaN         NaN         NaN  

[2401 rows x 24 columns]

It looks like there are a 2401 rows that are almost empty with a case number equal to zero. Lets confirm if these rows are all almost empty.

In [11]:
number_full_columns = data.apply(lambda x: x.count(), axis=1)

In [12]:
almost_empty_rows = number_full_columns[number_full_columns < 3].index
almost_empty_rows

Int64Index([ 6302,  6303,  6304,  6305,  6306,  6307,  6308,  6309,  6310,
             6311,
            ...
            25713, 25714, 25715, 25716, 25717, 25718, 25719, 25720, 25721,
            25722],
           dtype='int64', length=19421)

Not counting the 17020 empty rows, it looks like there are 2401 rows that just have one or two values. Looking at the idnexes, these are the same columns that have case number = 0 and almost no other data. These could be deleted. 

To make the rest of the analysis easier, it is best to delete the empty / almost empty rows now

In [13]:
# Drop rows that are empty or almost empty

data.drop(index=almost_empty_rows, inplace = True)

In [14]:
data

Case Number         Date    Year        Type             Country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex   ...         Species   \
0               Julie Wolfe    F  ...      White shark   
1           Adyson McNeely     F  ...              NaN   
2               John Denges    M  ...              NaN   
3                      male    M  ...        2 m shark   
4            Gustavo Ramos     M  ...  Tiger shark, 3m   
...                     ...  ...  ...              ...   
6297                   male    M  ...              NaN   
6298                  Ahmun    M  ...              NaN   
6299  Coast Guard personnel    M  ...              NaN   
6300        Jules Patterson    M  ...              NaN   
6301                   male    M  ...              NaN   

                     Investigator or Source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://s

Continuing withe the Year column, the format is not only weird (with a decimal point) but also no new information is given relative to the Date column, so Year could be deleted. However, seeing that date column has very diverse values, requiring extensive data cleaning, it is probably best to keep the year data converting it to int

Columns 'Type', 'Country', 'Area', 'Location', 'Activity' and 'Name' are giving relevant information not present elsewhere in the table and the format can only be a string. Sex could be more 'treatable' as a boolean, so lets see what the format is

In [15]:
type(data["Sex "][0])

str

Format is str, which could be changed to boolean, and we can also see that "sex " has a whitespace at the end, which should be deleted for higher clarity.

Lets continue by looking at the values of columns that are not shown when calling the df.head() function: 'Age', 'Injury', 'Fatal (Y/N)', 'Time'  

In [16]:
data.Age

0        57
1        11
2        48
3       NaN
4       NaN
       ... 
6297    NaN
6298    NaN
6299    NaN
6300    NaN
6301     15
Name: Age, Length: 6302, dtype: object

In [17]:
type(data.Age[1])

str

Age seems to be working but is returning a string. It would be useful to have an integer returned instead whenever the data is available

In [18]:
data.Injury

0       No injury to occupant, outrigger canoe and pad...
1                              Minor injury to left thigh
2            Injury to left lower leg from surfboard skeg
3                               Minor injury to lower leg
4       Lacerations to leg & hand shark PROVOKED INCIDENT
                              ...                        
6297                                                FATAL
6298                                                FATAL
6299                                                FATAL
6300                                                FATAL
6301    FATAL. "Shark bit him in half, carrying away t...
Name: Injury, Length: 6302, dtype: object

Injury seems to be giving somewhat valuable information in a string. It would not be easy to get visualizable data from this column but the information should be kept.

In [19]:
data["Fatal (Y/N)"]

0       N
1       N
2       N
3       N
4       N
       ..
6297    Y
6298    Y
6299    Y
6300    Y
6301    Y
Name: Fatal (Y/N), Length: 6302, dtype: object

In [20]:
type(data["Fatal (Y/N)"][1])

str

Fatal is giving valuable information stored as a string. While this works, a boolean could be valuable in this case

In [21]:
data.Time

0               18h00
1       14h00  -15h00
2               07h45
3                 NaN
4                 NaN
            ...      
6297              NaN
6298              NaN
6299              NaN
6300              NaN
6301              NaN
Name: Time, Length: 6302, dtype: object

In [22]:
type(data.Time[0])

str

While the column data seems to be giving valuable information, it is stored as a string, which is less usable, and there doesn't seem to be a consistent way of reporting the data. This could be done in a cleaner way

Species and investigator give relevant information that can only be stored as a string, so no changes seem necessary. The following three columns however, don't seem to offer any usable information. Lets look at them

In [23]:
data.pdf

0                 2018.06.25-Wolfe.pdf
1               2018.06.18-McNeely.pdf
2                2018.06.09-Denges.pdf
3             2018.06.08-Arrawarra.pdf
4                 2018.06.04-Ramos.pdf
                     ...              
6297            ND-0005-RoebuckBay.pdf
6298                 ND-0004-Ahmun.pdf
6299    ND-0003-Ocracoke_1900-1905.pdf
6300        ND-0002-JulesPatterson.pdf
6301                ND-0001-Ceylon.pdf
Name: pdf, Length: 6302, dtype: object

In [24]:
data["href formula"]

0       http://sharkattackfile.net/spreadsheets/pdf_di...
1       http://sharkattackfile.net/spreadsheets/pdf_di...
2       http://sharkattackfile.net/spreadsheets/pdf_di...
3       http://sharkattackfile.net/spreadsheets/pdf_di...
4       http://sharkattackfile.net/spreadsheets/pdf_di...
                              ...                        
6297    http://sharkattackfile.net/spreadsheets/pdf_di...
6298    http://sharkattackfile.net/spreadsheets/pdf_di...
6299    http://sharkattackfile.net/spreadsheets/pdf_di...
6300    http://sharkattackfile.net/spreadsheets/pdf_di...
6301    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href formula, Length: 6302, dtype: object

In [25]:
data.href

0       http://sharkattackfile.net/spreadsheets/pdf_di...
1       http://sharkattackfile.net/spreadsheets/pdf_di...
2       http://sharkattackfile.net/spreadsheets/pdf_di...
3       http://sharkattackfile.net/spreadsheets/pdf_di...
4       http://sharkattackfile.net/spreadsheets/pdf_di...
                              ...                        
6297    http://sharkattackfile.net/spreadsheets/pdf_di...
6298    http://sharkattackfile.net/spreadsheets/pdf_di...
6299    http://sharkattackfile.net/spreadsheets/pdf_di...
6300    http://sharkattackfile.net/spreadsheets/pdf_di...
6301    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href, Length: 6302, dtype: object

In [26]:
data["href formula"][0] == data.href[0]

True

In [27]:
data[data['href formula'] != data.href]

Case Number                  Date    Year          Type  \
50      2018.01.13           13-Jan-2018  2018.0    Unprovoked   
96      2017.08.29           29-Aug-2017  2017.0    Unprovoked   
131     2017.06.05            2017.06.05  2017.0    Unprovoked   
133     2017.06.11           11-Jun-2017  2017.0    Unprovoked   
141     2017.05.27           27-May-2017  2017.0       Invalid   
168     2017.03.19           19-Mar-2017  2017.0    Unprovoked   
234     2016.07.29           29-Jul-2016  2016.0    Unprovoked   
241   2016.07.23.a           23-Jul-2016  2016.0    Unprovoked   
276   2016.05.21.a           21-May-2016  2016.0    Unprovoked   
324   2015.12.21.a           21-Dec-2015  2015.0    Unprovoked   
332   2015.11.15.a           15-Nov-2015  2015.0    Unprovoked   
352   2015.10.05.a           05-Oct-2015  2015.0    Unprovoked   
401   2015.06.27.a           27-Jun-2015  2015.0    Unprovoked   
453     2015.01.24           24-Jan-2015  2015.0    Unprovoked   
484     2014.10.19           19-Oct-2014  2014.0       Boating   
485     2014.10.18           18-Oct-2014  2014.0    Unprovoked   
486     2014.10.17           17-Oct-2014  2014.0    Unprovoked   
493   2014.10.03.b           03-Oct-2014  2014.0       Boating   
494   2014.10.03.a           03-Oct-2014  2014.0       Boating   
515     2014.08.12           12-Aug-2014  2014.0    Unprovoked   
527     2014.07.22           22-Jul-2014  2014.0    Unprovoked   
545   2014.06.09.a           09-Jun-2014  2014.0    Unprovoked   
569     2014.04.03           03-Apr-2014  2014.0    Unprovoked   
583   2014.02.07.b           07-Feb-2014  2014.0      Provoked   
584     2014.02.07           07-Feb-2014  2014.0    Unprovoked   
585     2014.01.26           26-Jan-2014  2014.0      Provoked   
586     2014.01.25           25-Jan-2014  2014.0    Unprovoked   
587     2014.01.04           04-Jan-2014  2014.0  Sea Disaster   
588   2014.00.00.b                  2014  2014.0    Unprovoked   
590     2013.12.25           25-Dec-2013  2013.0    Unprovoked   
600     2013.11.22           22-Nov-2013  2013.0    Unprovoked   
601     2013.11.12           12-Nov-2013  2013.0    Unprovoked   
602     2013.11.10           10-Nov-2013  2013.0      Provoked   
603   2013.11.07.b           07-Nov-2013  2013.0    Unprovoked   
604   2013.11.07.a           07-Nov-2013  2013.0    Unprovoked   
605     2013.10.31           31-Oct-2013  2013.0    Unprovoked   
606     2013.10.28           28-Oct-2013  2013.0    Unprovoked   
607   2013.10.26.b           26-Oct-2013  2013.0    Unprovoked   
671   2013.05.27.b           27-May-2013  2013.0    Unprovoked   
672   2013.05.27.a           27-May-2013  2013.0    Unprovoked   
1046    2010.02.01           01-Feb-2010  2010.0      Provoked   
1441  2006.08.22.b           22-Aug-2006  2006.0       Invalid   
1442  2006.08.22.a           22-Aug-2006  2006.0       Invalid   
2702  1989.00.00.a                  1989  1989.0    Unprovoked   
3244    1975.01.19           19-Jan-1975  1975.0    Unprovoked   
3245  1975.00.00.b                  1975  1975.0    Unprovoked   
3775  1963.00.00.a                  1963  1963.0    Unprovoked   
3829  1962.03.25.b           25-Mar-1962  1962.0      Provoked   
4143    1958.08.31           31-Aug-1958  1958.0          Boat   
4154  1958.07.00.b              Jul-1958  1958.0    Unprovoked   
4621    1945.07.30           30-Jul-1945  1945.0  Sea Disaster   
4872    1936.08.04  Reported 04-Aug-1936  1936.0    Unprovoked   
4898    1935.08.24           24-Aug-1935  1935.0       Invalid   
4949  1934.01.08.R  Reported 08-Feb-1934  1934.0       Boating   
5245    1921.07.21           21-Jul-1921  1921.0    Unprovoked   
5349  1913.12.26.R  Reported 26-Dec-1913  1913.0  Sea Disaster   
5556    1900.09.05           05-Sep-1900  1900.0    Unprovoked   
5701    1890.06.26            06-26-1890  1890.0    Unprovoked   
6110  1819.07.08.R  Reported 08-Jul-1819  1819.0       Invalid   
6166  1637.00.00.R         Reported 1637  1637.

In [28]:
data[data['href formula'] != data.href].count()

Case Number               60
Date                      60
Year                      60
Type                      60
Country                   60
Area                      53
Location                  55
Activity                  56
Name                      59
Sex                       57
Age                       33
Injury                    59
Fatal (Y/N)               56
Time                      29
Species                   37
Investigator or Source    60
pdf                       60
href formula              59
href                      60
Case Number.1             60
Case Number.2             60
original order            60
Unnamed: 22                0
Unnamed: 23                0
dtype: int64

The pdf column seems to sotre a filename with the date of the attack and the victim's last name. This is repeated data and should be deleted. "href formula" and "href" store broken links (i.e. clicking on them takes you nowhere) to what seems to be a web page with data on shark attacks, so the data is irrelevant and should be deleted. Most of this links are even duplicated in both columns

Looking at columns "Case Number.1" and "Case Number.2" it looks like they could just be repeating the data on Case Number (and therefore Date)

In [29]:
data[data['Case Number.1'] != data['Case Number.2']]

Case Number                                  Date    Year  \
34      2018.04.03                           03-Apr-2018  2018.0   
117   2017.07.20.a                           20-Jul-2017  2017.0   
144     2017.05.06                           05-May-2017  2017.0   
217     2016.09.15                           16-Sep-2016  2016.0   
314   2016.01.24.b                           24-Jan-2016  2016.0   
334     2015.12.23                           07-Nov-2015  2015.0   
339   2015.10.28.a                           28-Oct-2015  2015.0   
560     2014.05.04                           04-May-2014  2014.0   
3522    1967.07.05                           05-Jul-1967  1967.0   
3795  1962,08.30.b                           30-Aug-1962  1962.0   
3880  1961.09.02.R                  Reported 06-Sep-1961  1961.0   
4403    1952.08.05                           05-Aug-1952  1952.0   
4425  1851.12.15.R                        \n1951.12.15.R  1951.0   
5275  1900.00.00.R  Reported to have taken place in 1919  1919.0   
5358  1913.07.10.R                  Reported 10-Jul-1913  1913.0   
5386  1911.07.31.R                  Reported 31-Jul-1911  1911.0   
5777  1884.04.38.R                  Reported 28-Apr-1884  1884.0   
5883  1876.06.18.R                 Reported 18-Jun- 1874  1874.0   
6009    1853,94.29                           29-Apr-1853  1853.0   
6122  1808.06.01.R                  Reported 01-May-1808  1808.0   

              Type       Country                   Area  \
34      Unprovoked  SOUTH AFRICA  Eastern Cape Province   
117     Unprovoked           USA             California   
144     Unprovoked        MEXICO    Baja California Sur   
217     Unprovoked     AUSTRALIA               Victoria   
314     Unprovoked           USA                  Texas   
334        Invalid           USA                Florida   
339     Unprovoked           USA                 Hawaii   
560     Unprovoked  SOUTH AFRICA  Western Cape Province   
3522    Unprovoked        TURKEY         Mugla Province   
3795          Boat        TURKEY       Antalya Province   
3880      Provoked         ITALY        Venice Province   
4403      Provoked         ITALY                 Teramo   
4425       Invalid    MOZAMBIQUE        Sofala Province   
5275       Boating         ITALY                    NaN   
5358    Unprovoked    COSTA RICA                    NaN   
5386    Unprovoked         SPAIN                Málaga    
5777    Unprovoked        BRAZIL                    NaN   
5883    Unprovoked        MEXICO              Veracruz    
6009    Unprovoked        GREECE                 Corfu    
6122  Sea Disaster       ANTIGUA                    NaN   

                                       Location  \
34                              St. Francis Bay   
117      Seal Rock, Goleta Beach, Santa Barbara   
144                   Los Arbolitos, Cabo Pulmo   
217                                 Bells Beach   
314                                Off Surfside   
334   Paradise Beach, Melbourne, Brevard County   
339                                Malaka, Oahu   
560                                  Simonstown   
3522                            Kucukada Island   
3795                                     Ucagiz   
3880                                   Chioggia   
4403                                 Giulianova   
4425                                      Beira   
5275                                     Savona   
5358                                 Port Limon   
5386                                      Ceuta   
5777                                      Bahia   
5883                     Tlacotalpan, Vera Cruz   
6009                                        NaN   
6122                           Falmouth Harbour   

                                     Activity  \
34                                    Surfing   
117                                       SUP   
144                                Snorkeling   
217                                   Surfing   
314                  

In [30]:
data[data['Case Number.1'] != data['Case Number.2']].count()

Case Number               20
Date                      20
Year                      20
Type                      20
Country                   20
Area                      16
Location                  19
Activity                  18
Name                      17
Sex                       17
Age                        8
Injury                    20
Fatal (Y/N)               17
Time                       8
Species                   11
Investigator or Source    20
pdf                       20
href formula              20
href                      20
Case Number.1             20
Case Number.2             20
original order            20
Unnamed: 22                0
Unnamed: 23                0
dtype: int64

In [31]:
data[data['Case Number.2'] != data['Case Number']]

Case Number                  Date    Year        Type    Country  \
390     2015.07-10           10-Jul-2015  2015.0  Unprovoked        USA   
4949  1934.01.08.R  Reported 08-Feb-1934  1934.0     Boating     TURKEY   
5488           NaN  Reported 06-Sep-1905  1905.0    Provoked        USA   
5944       1864.05              May-1864  1864.0  Unprovoked  MAURITIUS   

                             Area                         Location  \
390                    California  Huntington Beach, Orange County   
4949                     Istanbul       Haydarpasa jetty, Istanbul   
5488                      Florida     Fort Pierce, St Lucie County   
5944  300 miles east of Mauritius                              NaN   

            Activity                            Name Sex   ...  \
390          Surfing                    Danny Miskin    M  ...   
4949         Fishing                         2 males    M  ...   
5488         Fishing                     David Curry    M  ...   
5944  Fell overboard  male from the whaler Albatross    M  ...   

             Species              Investigator or Source  \
390   White shark, 7'                    KTLA, 7/10/2015   
4949              NaN                     C. Moore, GSAF   
5488              NaN     Muncie Evening Press, 8/6/1913   
5944              NaN  North Adams Transcript, 3/18/1898   

                           pdf  \
390      2015.07.10-Miskin.pdf   
4949   1924.02.08.R-Turkey.pdf   
5488    1905.09.06.R-Curry.pdf   
5944  1864.05.00.Mauritius.pdf   

                                           href formula  \
390   http://sharkattackfile.net/spreadsheets/pdf_di...   
4949  http://sharkattackfile.net/spreadsheets/pdf_di...   
5488  http://sharkattackfile.net/spreadsheets/pdf_di...   
5944  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
390   http://sharkattackfile.net/spreadsheets/pdf_di...    2015.07.10   
4949  http://sharkattackfile.net/spreadsheets/pdf_di...  1934.02.08.R   
5488  http://sharkattackfile.net/spreadsheets/pdf_di...  1905.09.06.R   
5944  http://sharkattackfile.net/spreadsheets/pdf_di...    1864.05.00   

     Case Number.2 original order Unnamed: 22 Unnamed: 23  
390     2015.07.10         5913.0         NaN         NaN  
4949  1934.02.08.R         1354.0         NaN         NaN  
5488  1905.09.06.R          815.0         NaN         NaN  
5944    1864.05.00          359.0         NaN         NaN  

[4 rows x 24 columns]

In [32]:
data[data['Case Number.2'] != data['Case Number']].count()

Case Number               3
Date                      4
Year                      4
Type                      4
Country                   4
Area                      4
Location                  3
Activity                  4
Name                      4
Sex                       4
Age                       1
Injury                    4
Fatal (Y/N)               4
Time                      1
Species                   1
Investigator or Source    4
pdf                       4
href formula              4
href                      4
Case Number.1             4
Case Number.2             4
original order            4
Unnamed: 22               0
Unnamed: 23               0
dtype: int64

It is clear that columns Case Number.1 and .2 offer no new information and should be deleted

In the last columns, we can see there is an 'original order' that could seems like it could serve as a unique identifier, but again it just seems like this data is just the inverse order of the date, so it really isn't giving much information and could also be removed.

In [33]:
data

Case Number         Date    Year        Type             Country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex   ...         Species   \
0               Julie Wolfe    F  ...      White shark   
1           Adyson McNeely     F  ...              NaN   
2               John Denges    M  ...              NaN   
3                      male    M  ...        2 m shark   
4            Gustavo Ramos     M  ...  Tiger shark, 3m   
...                     ...  ...  ...              ...   
6297                   male    M  ...              NaN   
6298                  Ahmun    M  ...              NaN   
6299  Coast Guard personnel    M  ...              NaN   
6300        Jules Patterson    M  ...              NaN   
6301                   male    M  ...              NaN   

                     Investigator or Source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://s

Finally, we can analyze columns like 'type', 'country', 'activity', 'sex', 'injury', 'fatal' and 'time' to see what unique values they have and if it is worth it to lose some data by grouping exceptional values into "other" categories 

In [34]:
data.Type.value_counts()

Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          203
Boat             137
Questionable       2
Boatomg            1
Name: Type, dtype: int64

The type column has just 8 different values, which is great for working with the data. however, it seems that there is one value 'Boatomg' that should be 'Boating', as well as a value 'Boat' that should be 'Boating'

In [35]:
data.Country.value_counts()

USA                 2229
AUSTRALIA           1338
SOUTH AFRICA         579
PAPUA NEW GUINEA     134
NEW ZEALAND          128
                    ... 
BAY OF BENGAL          1
JAVA                   1
SUDAN?                 1
Sierra Leone           1
IRELAND                1
Name: Country, Length: 212, dtype: int64

Country, however, has 212 different values, which is more than the number of countries in the world, so it is probably better to stick to the largest countries

In [36]:
data.Activity.value_counts().head(15)

Surfing          971
Swimming         869
Fishing          431
Spearfishing     333
Bathing          162
Wading           149
Diving           127
Standing          99
Snorkeling        89
Scuba diving      76
Body boarding     61
Body surfing      49
Swimming          47
Kayaking          33
Pearl diving      32
Name: Activity, dtype: int64

Similarly, Activity has fifteen hundred different values, so it is better to stick to a top X. Some of the most relevant ones seem to be refering to similar things (eg. Swimming and Bathing) and could be combined

In [37]:
data['Sex '].value_counts()

M      5094
F       637
N         2
M         2
lli       1
.         1
Name: Sex , dtype: int64

Sex has an insignificant amount of values that have an error

In [38]:
data.Injury.value_counts().head

<bound method NDFrame.head of FATAL                                                                                     802
Survived                                                                                   97
Foot bitten                                                                                87
No injury                                                                                  82
Leg bitten                                                                                 72
                                                                                         ... 
FATAL, thigh and abdomen bitten                                                             1
FATAL, right leg bitten thigh to calf                                                       1
Lacerations & puncture wounds  to left thigh                                                1
FATAL, right arm severed, chest punctured                                                   1
Gietel grabbed the shark's tai

Again, injury has more than three thousand different values, os sticking with the top X seems best

In [39]:
data['Fatal (Y/N)'].value_counts()

N          4293
Y          1388
UNKNOWN      71
 N            7
2017          1
y             1
M             1
N             1
Name: Fatal (Y/N), dtype: int64

Fatal seems to be quite clean, with just some exceptions not working as they should and most can be cleaned by removing spaces

In [40]:
data.Time.value_counts()

Afternoon        187
11h00            128
Morning          121
12h00            109
15h00            108
                ... 
07h32              1
07h10              1
>17h30             1
14h30 / 15h30      1
13h06              1
Name: Time, Length: 366, dtype: int64

The time column seems to have all sorts of different values, so it will not be the best column to use without exhaustive data cleaning

## Treat dataset

First, we can delete the columns that give repeated / non-relevant information

In [41]:
# Drop columns that are full of Nan, repeated or not relevant

columns_to_drop = ["Case Number", "pdf", "href formula", "href", "Case Number.1", "Case Number.2", "original order", "Unnamed: 22", "Unnamed: 23"]

data.drop(columns=columns_to_drop, inplace = True)

In [42]:
data

Date    Year        Type             Country               Area  \
0     25-Jun-2018  2018.0     Boating                 USA         California   
1     18-Jun-2018  2018.0  Unprovoked                 USA            Georgia   
2     09-Jun-2018  2018.0     Invalid                 USA             Hawaii   
3     08-Jun-2018  2018.0  Unprovoked           AUSTRALIA    New South Wales   
4     04-Jun-2018  2018.0    Provoked              MEXICO             Colima   
...           ...     ...         ...                 ...                ...   
6297  Before 1903     0.0  Unprovoked           AUSTRALIA  Western Australia   
6298  Before 1903     0.0  Unprovoked           AUSTRALIA  Western Australia   
6299    1900-1905     0.0  Unprovoked                 USA     North Carolina   
6300    1883-1889     0.0  Unprovoked              PANAMA                NaN   
6301    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   Eastern Province   

                                 Location      Activity  \
0             Oceanside, San Diego County      Paddling   
1          St. Simon Island, Glynn County      Standing   
2                            Habush, Oahu       Surfing   
3                      Arrawarra Headland       Surfing   
4                                La Ticla   Free diving   
...                                   ...           ...   
6297                          Roebuck Bay        Diving   
6298                                  NaN  Pearl diving   
6299                       Ocracoke Inlet      Swimming   
6300                 Panama Bay 8ºN, 79ºW           NaN   
6301  Below the English fort, Trincomalee      Swimming   

                       Name Sex   Age  \
0               Julie Wolfe    F   57   
1           Adyson McNeely     F   11   
2               John Denges    M   48   
3                      male    M  NaN   
4            Gustavo Ramos     M  NaN   
...                     ...  ...  ...   
6297                   male    M  NaN   
6298                  Ahmun    M  NaN   
6299  Coast Guard personnel    M  NaN   
6300        Jules Patterson    M  NaN   
6301                   male    M   15   

                                                 Injury Fatal (Y/N)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6297                                              FATAL           Y   
6298                                              FATAL           Y   
6299                                              FATAL           Y   
6300                                              FATAL           Y   
6301  FATAL. "Shark bit him in half, carrying away t...           Y   

               Time         Species                  Investigator or Source  
0             18h00      White shark                       R. Collier, GSAF  
1     14h00  -15h00              NaN         K.McMurray, TrackingSharks.com  
2             07h45              NaN         K.McMurray, TrackingSharks.com  
3               NaN        2 m shark                         B. Myatt, GSAF  
4               NaN  Tiger shark, 3m                              A .Kipper  
...             ...              ...                                    ...  
6297            NaN              NaN       H. Taunton; N. Bartlett,  p. 234  
6298            NaN              NaN  H. Taunton; N. Bartlett,  pp. 233-234  
6299            NaN              NaN   F. Schwartz, p.23; C. Creswell, GSAF  
6300            NaN              NaN                    The Sun, 10/20/1938  
6301            NaN              NaN                             S.W. Baker  

[6302 rows x 15 columns]

At this point, we have gone from an original dataframe of 25723 rows × 24 columns to one with 6302 rows × 14 columns, keeping all the relevant information

Next we can remove extra spaces from column names, change spaces between words to underscores and have all column names be in lower case, to make it easier to call each column

In [43]:
#to remove trailing spaces at the begining and end of the column names, spaces between words to undersocres and change to lower case

data.columns = data.columns.str.lower().str.strip().str.replace(' ','_').str.replace('/','-')

In [44]:
data

date    year        type             country               area  \
0     25-Jun-2018  2018.0     Boating                 USA         California   
1     18-Jun-2018  2018.0  Unprovoked                 USA            Georgia   
2     09-Jun-2018  2018.0     Invalid                 USA             Hawaii   
3     08-Jun-2018  2018.0  Unprovoked           AUSTRALIA    New South Wales   
4     04-Jun-2018  2018.0    Provoked              MEXICO             Colima   
...           ...     ...         ...                 ...                ...   
6297  Before 1903     0.0  Unprovoked           AUSTRALIA  Western Australia   
6298  Before 1903     0.0  Unprovoked           AUSTRALIA  Western Australia   
6299    1900-1905     0.0  Unprovoked                 USA     North Carolina   
6300    1883-1889     0.0  Unprovoked              PANAMA                NaN   
6301    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   Eastern Province   

                                 location      activity  \
0             Oceanside, San Diego County      Paddling   
1          St. Simon Island, Glynn County      Standing   
2                            Habush, Oahu       Surfing   
3                      Arrawarra Headland       Surfing   
4                                La Ticla   Free diving   
...                                   ...           ...   
6297                          Roebuck Bay        Diving   
6298                                  NaN  Pearl diving   
6299                       Ocracoke Inlet      Swimming   
6300                 Panama Bay 8ºN, 79ºW           NaN   
6301  Below the English fort, Trincomalee      Swimming   

                       name sex  age  \
0               Julie Wolfe   F   57   
1           Adyson McNeely    F   11   
2               John Denges   M   48   
3                      male   M  NaN   
4            Gustavo Ramos    M  NaN   
...                     ...  ..  ...   
6297                   male   M  NaN   
6298                  Ahmun   M  NaN   
6299  Coast Guard personnel   M  NaN   
6300        Jules Patterson   M  NaN   
6301                   male   M   15   

                                                 injury fatal_(y-n)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6297                                              FATAL           Y   
6298                                              FATAL           Y   
6299                                              FATAL           Y   
6300                                              FATAL           Y   
6301  FATAL. "Shark bit him in half, carrying away t...           Y   

               time          species                 investigator_or_source  
0             18h00      White shark                       R. Collier, GSAF  
1     14h00  -15h00              NaN         K.McMurray, TrackingSharks.com  
2             07h45              NaN         K.McMurray, TrackingSharks.com  
3               NaN        2 m shark                         B. Myatt, GSAF  
4               NaN  Tiger shark, 3m                              A .Kipper  
...             ...              ...                                    ...  
6297            NaN              NaN       H. Taunton; N. Bartlett,  p. 234  
6298            NaN              NaN  H. Taunton; N. Bartlett,  pp. 233-234  
6299            NaN              NaN   F. Schwartz, p.23; C. Creswell, GSAF  
6300            NaN              NaN                    The Sun, 10/20/1938  
6301            NaN              NaN                             S.W. Baker  

[6302 rows x 15 columns]

Next, lets change the country names to a 'capitalize' case

In [45]:
# convert country case to 'capitalize'

data['country'] = data['country'].str.capitalize()

Now lets covert data formats from all strings to more useful formats

In [46]:
# convert year to int, if there is an error (probably NaN), return zero and we can later ignore that value

def convert_float_to_int(flt):
    try:
        return int(flt)
    except:
        return 0
    
data['year'] = data['year'].apply(convert_float_to_int)

In [47]:
# convert age to int, if there is an error (probably NaN), return zero and we can later ignore that value

def convert_str_to_int(string):
    try:
        return int(string)
    except:
        return 0

data['age'] = data['age'].apply(convert_str_to_int)

In [48]:
data.head()

date  year        type    country             area  \
0  25-Jun-2018  2018     Boating        Usa       California   
1  18-Jun-2018  2018  Unprovoked        Usa          Georgia   
2  09-Jun-2018  2018     Invalid        Usa           Hawaii   
3  08-Jun-2018  2018  Unprovoked  Australia  New South Wales   
4  04-Jun-2018  2018    Provoked     Mexico           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M    0   
4                        La Ticla  Free diving   Gustavo Ramos    M    0   

                                              injury fatal_(y-n)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            time          species          investigator_or_source  
0          18h00      White shark                R. Collier, GSAF  
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com  
2          07h45              NaN  K.McMurray, TrackingSharks.com  
3            NaN        2 m shark                  B. Myatt, GSAF  
4            NaN  Tiger shark, 3m                       A .Kipper

We can now clean the 'fatal' column to remove extra spaces

In [49]:
data['fatal_(y-n)'] = data['fatal_(y-n)'].str.strip()

In [50]:
data['fatal_(y-n)'].value_counts()

N          4301
Y          1388
UNKNOWN      71
2017          1
y             1
M             1
Name: fatal_(y-n), dtype: int64

Now we can clean the remaining 'weird' results

In [51]:
data['fatal_(y-n)'] = data['fatal_(y-n)'].str.replace('y','Y').str.replace('M','N').str.replace('2017','UNKNOWN')

In [52]:
data['fatal_(y-n)'].value_counts()

N          4302
Y          1389
UNKNOWN      72
Name: fatal_(y-n), dtype: int64

In the activity column we can also do some cleaning

In [53]:
# remove white spaces

data.activity = data.activity.str.strip()

In [54]:
data.activity.value_counts().head(15)

Surfing           976
Swimming          916
Fishing           444
Spearfishing      340
Bathing           165
Wading            149
Diving            130
Standing          101
Snorkeling         89
Scuba diving       77
Body boarding      62
Body surfing       50
Kayaking           35
Fell overboard     33
Treading water     32
Name: activity, dtype: int64

In [55]:
# Using regex to combine bathing and swimming

data.activity = data.activity.apply(lambda x: re.sub(r'^(Bathing)$', 'Swimming' , str(x)))

In [56]:
# Using regex to combine fishing and spearfishing

data.activity = data.activity.apply(lambda x: re.sub(r'^(Spearfishing)$', 'Fishing' , str(x)))

In [57]:
# Using regex to combine wading and standing

data.activity = data.activity.apply(lambda x: re.sub(r'^(Standing)$', 'Wading' , str(x)))

In [58]:
data.activity.value_counts().head(15)

Swimming          1081
Surfing            976
Fishing            784
nan                544
Wading             250
Diving             130
Snorkeling          89
Scuba diving        77
Body boarding       62
Body surfing        50
Kayaking            35
Fell overboard      33
Treading water      32
Pearl diving        32
Free diving         30
Name: activity, dtype: int64

In [59]:
# Using regex to combine free diving and pearl diving

data.activity = data.activity.apply(lambda x: re.sub(r'^(Pearl diving)$', 'Free diving' , str(x)))

A small fix is changing the 'Boatomg' and 'Boat' values in column types to 'Boating'

In [60]:
data['type'] = data['type'].str.replace('Boatomg', 'Boating')
data['type'] = data['type'].str.replace('Boat', 'Boating')

# Replace again to replace 'boatinging' to 'boating'
data['type'] = data['type'].str.replace('Boatinging', 'Boating')

The last step is to save the datafram to a csv file so it can be used in a different notebook

In [61]:
data.to_csv('clean-shark-data.csv')

We are now ready to test the hypotheses that were described at the start of the notebook